### PR TITLE
Add support for comparators to Sort.chpl

### PR DESCRIPTION
**Pull request by ben-albrecht**
### Comparators

In Chapel, comparators  are passed to the sort function as records with `.key(a)` or `.compare(a, b)` methods, rather than as functions themselves.

Some examples from the Sort.chpl *chpldoc-umentation*:

Example of `key(a)`:

```chapel

  var Array = [-1, -4, 2, 3];

  // Empty record serves as comparator
  record comparator { }

  // key method maps an element to the value to be used for comparison
  proc comparator.key(a) { return abs(a); }

  myComparator = new comparator();

  sort(Array, comparator=myComparator);

  // This will output: -1, 2, 3, -4
  writeln(Array);
```

Example of `compare(a, b)` sorting by absolute values:

```chapel

  var Array = [-1, -4, 2, 3];

  // Empty record serves as comparator
  record comparator { }

  // compare method defines how 2 elements are compared
  proc comparator.compare(a, b) {
      return abs(a) - abs(b);
  }

  myComparator = new comparator();

  sort(Array, comparator=myComparator);

  // This will output: -1, 2, 3, -4
  writeln(Array);
```

### Other stuff
* Added thorough chpldocumentation with examples
* Added tests for comparator support and expected errors (runtime and compile-time).
* Began deprecation of old sort function interface (supported `reverse` and `doublecheck` arguments + PascalCase instead of camelCase)
    * Threw `pragma "no doc"` on old sort functions
    * Staged deprecation warning as a comment
        * This will be uncommented in another PR in which we may decide to implement a standardized deprecation message function.
* Updated the base `chpl_compare` method to utilize comparators, and rely on the `defaultComparator` when no comparators are passed.
* Added convenient sort interface `sort()`, which currently acts as a wrapper for `quickSort`.
    * Updated standard library calls to `QuickSort` to `sort`

### Future Work
* Fully deprecate old sort interfaces
    *  Suppressable deprecation warnings for 1.14, then removed for 1.15
* Add comparator support to `selectionSort`
    * Work required with user-defined 
* Implement radix sort
* Possibly move individual sort algorithms to a SortAlgorithms module or submodule to reduce the complexity of Sort (since most uses of this library will only want to call 'sort', 'sorted' or 'isSorted').
* implement parallel sorting algorithms and have some reasonable way for users to request serial or parallel sorts.
* Create performance tracking tests and optimize/parallelize sort functions further
* Move Sort tests from `test/modules/standard` -> `test/modules/packages`

### TODO:

* [x] Add more complete testing (reverse, doublecheck, old version vs. new version)
* [x] Complete *chpldoc-umentation*
* [x] Rerun paratest
   * `[Summary: #Successes = 6260 | #Failures = 0 | #Futures = 0]`

### CODE REVIEW

File: modules/packages/Sort.chpl [hunk 1]
```diff
@@ -18,27 +18,178 @@
...
+default comparison of array elements. To use a comparator, define a
+record with either a ``key(a)`` or ``compare(a, b)`` method, and pass an
+instance of that record to the sort function.
+
+Most sort routines support comparators, which is denoted by their function
+signature.
+
+The ``key(a)`` method accepts one argument with the array ``eltType`` and
+returns a value that supports all of comparison operations: ``<``, ``<=``,
+``>``, ``>=``, or ``==``.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: Are all of these actually used? E.g. you could implement them all with < ... if one didn't resolve.

File: modules/packages/Sort.chpl [hunk 2]
```diff
@@ -18,27 +18,178 @@
...
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+
+The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
+returns an integer between ``1`` if ``a > b``, ``0`` if ``a == b``, and ``-1`` if ``a < b``.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'd reword this comment. Also, generally speaking such comparison functions allow values >1 or < -1. See man qsort, which says:

        The comparison function must return an integer less than, equal to, or
     greater than zero if the first argument is considered to be respectively
     less than, equal to, or greater than the second.

I'd write:
The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
returns an integer indicating how a and b compare to each other. It should return

  * > 0 if ``a > b``
  *    0 if ``a == b``
  * < 0 if ``a < b`

Comment (mppf) [inline modules/packages/Sort.chpl]: By the way, the reason that it's not just -1 0 and 1 is so that you can write this:

    proc compare(a:int, b:int) {
      return a - b;
    }

File: modules/packages/Sort.chpl [hunk 3]
```diff
@@ -18,27 +18,178 @@
...
+
+  // compare method defines how 2 elements are compared
+  proc comparator.compare(a, b) {
+    if abs(a) > abs(b) {
+      return 1;
+    } else if abs(a) < abs(b) {
+      return -1;
+    } else {
+      return 0;
+    }
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'd recommend you write this function body as

    return abs(a) - abs(b)

to emphasize and demonstrate the point about > 1 or < -1

File: modules/packages/Sort.chpl [hunk 4]
```diff
@@ -18,27 +18,178 @@
...
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: Hmm... interesting. Could you mark this as "current behavior" somehow? Once we add radix sort (which I think is ... key ... heh) it will not be able to use compare at all.

E.g.

In the current implementation, if both...

File: modules/packages/Sort.chpl [hunk 5]
```diff
@@ -18,27 +18,178 @@
...
+  use Reflection;
+
+  // TODO -- we could also check types
+  // Use comparator.compare(a, b) if is defined by user
+  if canResolveMethod(comparator, "compare", a, b) {
+    if reverse { return -comparator.compare(a, b); }
+    else { return  comparator.compare(a, b); }
+  }
+
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'm fine with this comment, but Knuth volume 3 has a good discussion. You generally will only benefit from such a transformation if the values being sorted are much larger than their keys. I actually think that's less likely than the value being close to the same size as the key.

File: modules/packages/Sort.chpl [hunk 6]
```diff
@@ -18,27 +18,178 @@
...
+      1 if a > b
+      0 if a == b
+     -1 if a < b
+
+ */
+private inline proc chpl_compare(a, b, param reverse=false, comparator:?rec=emptyRecord): int {
+  use Reflection;
+
+  // TODO -- we could also check types
+  // Use comparator.compare(a, b) if is defined by user
```
Comment (mppf) [inline modules/packages/Sort.chpl]: Do you know if Reflection works with type methods? I'm not so sure... I'd expect we'd need to beef up the Reflection module in order to check for a type method.

File: modules/packages/Sort.chpl (hunk 1 above)
Comment (mppf) [inline modules/packages/Sort.chpl]: Actually I think it would be better if you said that the value returned by `key` must support all numeric operations, including comparison ( which you listed), and also things like - and <<.

I think it's more efficient to use - for our default comparator...
I'd use << and & for radix sorting...

File: modules/packages/Sort.chpl [hunk 7]
```diff
@@ -18,27 +18,178 @@
...
+  if canResolveMethod(comparator, "compare", a, b) {
+    if reverse { return -comparator.compare(a, b); }
+    else { return  comparator.compare(a, b); }
+  }
+
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
+  // Compare results of comparator.key(a) if is defined by user
+  if canResolveMethod(comparator, "key", a) && canResolveMethod(comparator, "key", b) then
+    return chpl_compare(comparator.key(a), comparator.key(b), reverse);
```
Comment (mppf) [inline modules/packages/Sort.chpl]: See my comment above. I think this should return

if reverse then comparator.key(b) - comparator.key(a);
else comparator.key(a) - comparator.key(b);

I don't think we're getting much out of the `key` interface if it does not return a numeric value. I think people who have a "key" that is not numeric should just use the .compare interface.

Comment (mppf) [inline modules/packages/Sort.chpl]: Of course you might disagree if you wanted .key to support strings, that's an interesting use...

File: modules/packages/Sort.chpl [hunk 8]
```diff
@@ -18,27 +18,178 @@
...
+
+  if reverse {
+    if a < b { return 1; }
+    else if b < a { return -1; }
   } else {
-    if reverse then return a > b;
-    else return a < b;
+    if a < b { return -1; }
+    else if b < a { return 1; }
```
Comment (mppf) [inline modules/packages/Sort.chpl]: So, in the end you only used < if neither .key nor .compare were available? When does this happen? It probably needs mention in the module-level comment above.

Comment (mppf) [inline modules/packages/Sort.chpl]: Or else this code can be removed?

File: modules/packages/Sort.chpl [hunk 9]
```diff
@@ -18,27 +18,178 @@
...
+    :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+private proc chpl_check_comparator(comparator, a) {
+  use Reflection;
+
+  if comparator.type == defaultComparator {}
+  else if canResolveMethod(comparator, "compare", a, a) {}
+  else if canResolveMethod(comparator, "key", a) {}
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I think it would make sense to do some checks on procs available on the return type of key here as well.

File: modules/packages/Sort.chpl [hunk 10]
```diff
@@ -47,24 +198,27 @@ private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
    :type doublecheck: `bool`
    :arg reverse: Sort in reverse numerical order
    :type reverse: `bool`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
 
  */
-proc BubbleSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false) where Dom.rank == 1 {
+proc BubbleSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false, comparator:?rec=emptyRecord) where Dom.rank == 1 && isRecord(rec) {
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I personally think that this `reverse` business is a waste of arguments.
Also doublecheck.

I can see keeping it for backwards compatibility, but I'd deprecate it.
For that reason I might prefer the actual sort algorithms be implemented with comparator and you just make other procs with the reverse argument that don't take in a comparator.
E.g. remove reverse and doublecheck from this one, and add
proc BubbleSort(Data, doublecheck=false, param reverse=false)
for backwards compatibility.

I'd probably criticize anybody who actually *calls* BubbleSort though. I'd personally advocate that we jettison it from the library completely.

But my comments will probably apply to the other sort variants.

File: modules/packages/Sort.chpl [hunk 11]
```diff
@@ -163,46 +323,49 @@ proc InsertionSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false
...
   var A2 = Data[mid+1..hi];
   cobegin {
-    { _MergeSort(A1, minlen, reverse=reverse); }
-    { _MergeSort(A2, minlen, reverse=reverse); }
+    { _MergeSort(A1, minlen, reverse=reverse, comparator=comparator); }
+    { _MergeSort(A2, minlen, reverse=reverse, comparator=comparator); }
   }
 
-  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse)) do a = _a;
+  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse, comparator=comparator)) do a = _a;
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I have trouble believing that this iterator-based implementation of MergeSort on arrays is anything other than a test of a language feature.
I'd expect MergeSort to outperform QuickSort if implemented well. I think this iterator is just adding overhead. It would be nice to add a comment indicating it's adding overhead (iterators with more than 1 yield can't be inlined) at not much benefit since the iterator is only called once.

File: modules/packages/Sort.chpl (hunk 10 above)
Comment (mppf) [inline modules/packages/Sort.chpl]: Also, I wouldn't put isRecord(rec) in the where clause. Let it be an error if chpl_check_comparator doesn't work. Nothing about how you are using the comparator needs it to be a record (it could be a class, or maybe somebody juts made methods on int, etc)

File: modules/packages/Sort.chpl [hunk 12]
```diff
@@ -263,14 +433,15 @@ proc QuickSort(Data: [?Dom] ?eltType, minlen=16, doublecheck=false, param revers
   Data(loptr) = pivotVal;
 
   //  cobegin {
```
Comment (mppf) [inline modules/packages/Sort.chpl]: Why is this Cobegin commented out?
I think it would be worthwhile to allow the user to choose whether the sort is parallel.

File: modules/packages/Sort.chpl [hunk 13]
```diff
@@ -306,11 +477,14 @@ proc SelectionSort(Data: [?Dom], doublecheck=false, param reverse=false) where D
    :type str: `string`
    :arg reverse: if true, expect the values to be sorted in reverse.
    :type reverse: `bool`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
 
  */
-inline proc VerifySort(Data: [?Dom] ?eltType, str: string, param reverse=false) {
+inline proc VerifySort(Data: [?Dom] ?eltType, str: string, param reverse=false, comparator:?rec=emptyRecord) where isRecord(rec) {
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'm very sure that this method does not need to be inline.

File: modules/packages/Sort.chpl [hunk 14]
```diff
@@ -18,27 +18,178 @@
...
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
 
  */
+
```
Comment (mppf) [inline modules/packages/Sort.chpl]: At the top-level, our Sort module is providing the wrong interface. We ask users to make choices that should be handled by our default.

For that reason, I'd recommend:
 * creating compatibility versions that use a special comparator and have existing signatures (and can be deprecated later)
 * removing reverse and  doublecheck arguments from the "real" versions of each sort algorithm
 * Also, for the "real" version of each, please rename it to not be upper case. This is one of the only places in our module where procs are not following the dominant style of starting with a lower case. That has always bothered me.
 * promoting VerifySort to a top-level function and rename it. I would call it isOrdered. As with the new sort routines, it should not take in a reverse argument - just a mapper. Make a deprecated version with the old signature if it's actually called in any tests.
 * Finally, we need to provide a way for users to get a parallel and non-parallel sort. We need a "Sort" method that chooses our best implementation. I would recommend we call this "sort" or "comparisonSort" and expect that later we will add "radixSort" which will have slightly different requirements on the comparator. I'd recommend using a non-param argument to indicate whether or not the sort should be parallel. A reasonable alternative would be to use a different method name for parallel/not parallel sorts.

File: modules/packages/Sort.chpl (hunk 4 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: We can swap the default priority for all sort functions. compare > key order was chosen somewhat arbitrarily.

File: modules/packages/Sort.chpl (hunk 7 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: Your proposed change is much more elegant, but I would like to keep .key general if possible, e.g. supporting strings. However, a difference in performance might persuade me one way over the other.

File: modules/packages/Sort.chpl (hunk 8 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: This is the default behavior when a comparator is not passed. If a user passes a comparator without a correctly defined `.key(a)` or `.compare(a,b)` method, they get a compile error.

File: modules/packages/Sort.chpl (hunk 10 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: * :+1: to deprecating `reverse` and `doublecheck` - Probably worth getting input from the rest of the team on this.
* I like bubble sort in that it's like the *hello world* of sorting. I feel like it's present more for completeness rather than actual usefulness.

File: modules/packages/Sort.chpl (hunk 6 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: Based on testing, they appear to be working fine.

File: modules/packages/Sort.chpl (hunk 11 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: Agreed, but I'm going to leave this change for a separate PR for now.

File: modules/packages/Sort.chpl (hunk 12 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: Not sure, this is how I found it... I will leave this change to another PR as well.

File: modules/packages/Sort.chpl (hunk 14 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: I will take care of all of these, except for the last point. I'd like to split that change off into a separate more broad "Sort Improvements" PR.

**Comment by ben-albrecht** [comment]
@mppf - I think this is ready for final review now. Some comments:

* I've reverted the `compare` method back to the original basic implementation. I think we can optimize this in the future with overloaded version for different argument types.
* I've also decided to keep the `ReverseComparator` relatively simple for now. We can build upon this implementation for radixSort support in the future.
* The deprecated functions have been `pragma "no doc"`'d and I've added a commented-out deprecation message. I will un-comment this deprecation message and fix all the tests it breaks in a follow-up PR.
    * Maybe it's time to add a standardized deprecation message with a compiler flag for silencing the warning.
* `selectionSort` comparator support will also come in the near future. The challenge here lies in properly defining a comparator-supported reduction class.

File: modules/internal/DefaultAssociative.chpl [hunk 15]
```diff
@@ -407,7 +407,7 @@ module DefaultAssociative {
       for (tmp, slot) in zip(tableCopy.domain, _fullSlots()) do
         tableCopy(tmp) = table[slot].idx;
   
-      QuickSort(tableCopy);
+      quickSort(tableCopy);
```
Comment (mppf) [inline modules/internal/DefaultAssociative.chpl]: quickSort -> sort please

File: modules/internal/DefaultAssociative.chpl [hunk 16]
```diff
@@ -676,7 +676,7 @@ module DefaultAssociative {
       for (copy, slot) in zip(tableCopy.domain, dom._fullSlots()) do
         tableCopy(copy) = data(slot);
   
-      QuickSort(tableCopy);
+      quickSort(tableCopy);
```
Comment (mppf) [inline modules/internal/DefaultAssociative.chpl]: here too

File: modules/packages/Sort.chpl [hunk 17]
```diff
@@ -18,53 +18,321 @@
...
+  // Empty record serves as comparator
+  record Comparator { }
+
+  // key method maps an element to the value to be used for comparison
+  proc Comparator.key(a) { return abs(a); }
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
```
Comment (mppf) [inline modules/packages/Sort.chpl]: QuickSort -> sort here too please

File: modules/packages/Sort.chpl [hunk 18]
```diff
@@ -18,53 +18,321 @@
...
+  record Comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc Comparator.compare(a, b) {
+    return abs(a) - abs(b);
+  }
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
```
Comment (mppf) [inline modules/packages/Sort.chpl]: QuickSort -> sort

File: modules/packages/Sort.chpl [hunk 19]
```diff
@@ -18,53 +18,321 @@
...
+
+A module-defined :const:`reversecomparator` can be passed to a sort function to
+reverse the default sorting order.
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Using module-defined 'reversecomparator'
+  QuickSort(Array, comprator=reversecomparator)
```
Comment (mppf) [inline modules/packages/Sort.chpl]: QuickSort -> sort

File: modules/packages/Sort.chpl [hunk 20]
```diff
@@ -18,53 +18,321 @@
...
+  writeln(Array);
+
+Reverse Comparator
+~~~~~~~~~~~~~~~~~~
+
+Sort functions in Chapel do not have a ``reverse`` argument. Instead, reverse
+sorting is handled through the comparator interface.
+
+A module-defined :const:`reversecomparator` can be passed to a sort function to
+reverse the default sorting order.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I would have expected it to be named reverseComparator. Is there a reason you opted for the all-lower-case version?

File: modules/packages/Sort.chpl [hunk 21]
```diff
@@ -18,53 +18,321 @@
...
+  record Comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc Comparator.compare(a, b) {
+    return abs(a) - abs(b);
+  }
+
+  var absReverseComparator: ReverseComparator(Comparator);
+
+  QuickSort(Array, comparator=absReverseComparator);
```
Comment (mppf) [inline modules/packages/Sort.chpl]: QuickSort -> sort

File: modules/packages/Sort.chpl [hunk 22]
```diff
@@ -18,53 +18,321 @@
...
-    if reverse then return a >= b;
-    else return a <= b;
+
+/* Module-defined comparators */
+
+/*
+  Instance of :record:`DefaultComparator` used as default ``comparator=``
+  argument when no comparator is passed to a sort function
+*/
+const defaultcomparator: DefaultComparator;
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I think I would call this 'defaultOrder' or 'defaultComparator'

File: modules/packages/Sort.chpl [hunk 23]
```diff
@@ -18,53 +18,321 @@
...
+
+
+
+/*
+   Instance of :record:`ReverseComparator`. Pass this as the ``comparator=``
+   argument of a sort function to reverse the sort order.
+ */
+const reversecomparator: ReverseComparator(DefaultComparator);
```
Comment (mppf) [inline modules/packages/Sort.chpl]: Why not 'reverseComparator' ?
'reverseOrder' is another possible name for this, as is 'reverseDefaultComparator'.

File: modules/packages/Sort.chpl [hunk 24]
```diff
@@ -18,53 +18,321 @@
...
+    // Check return type of compare
+    type comparetype = comparator.compare(data, data).type;
+    if !(isNumericType(comparetype)) {
+      compilerError("The compare method must return a numeric type");
+    }
+  }
+  else if canResolveMethod(comparator, "key", data) {
+    // Check return type of key
+    type keytype = comparator.key(data).type;
+    if !(isNumericType(keytype) || isStringType(keytype)) {
```
Comment (mppf) [inline modules/packages/Sort.chpl]: The comments above indicate 'key must support numeric operations'. Should it say 'key must support numeric operations or be a string' ?

File: modules/packages/Sort.chpl [hunk 25]
```diff
@@ -18,53 +18,321 @@
...
+    // If we make it this far, the passed comparator was defined incorrectly
+    compilerError("The comparator record requires a 'key(a)' or 'compare(a, b)' method");
+  }
+}
+
+
+/* Basic Functions */
+
+/*
+   Convenience wrapper around :proc:`quickSort`.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'd prefer we described this function differently. Could it say

General purpose sorting interface. 

.. note:

  This method currently calls a sequential quickSort but it will call faster algorithms once they are implemented.

File: modules/packages/Sort.chpl [hunk 26]
```diff
@@ -18,53 +18,321 @@
...
+   :arg Data: The array to verify
+   :type Data: [] `eltType`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+   :returns: ``true`` if array is sorted
+   :rtype: `bool`
+ */
+proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator): bool {
+  chpl_check_comparator(comparator, eltType);
+  for i in Dom.low..Dom.high-1 do
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I think it would be fun to write this in parallel but it doesn't need to happen right now...

File: modules/packages/Sort.chpl [hunk 27]
```diff
@@ -18,53 +18,321 @@
...
+// This is a first draft "sorterator" which is designed to take some
+// other iterator/iterable and yield its elements, in sorted order.
+//
+// The main limitations in the current code are (1) it should put some
+// sort of constraint on 'x' to limit it to types for which this makes
+// sense; and (2) there should be some generic way to say "y is an
+// array of x's element type" (or to infer its element type) without
+// saying a priori how big it is.  Without these mods, the result is
+// that the sorterator works when it does and probably is confusing
+// when it doesn't.
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'm confused by this comment. In the code below, y has element type inferred from x, and there is no mention of its size.

File: modules/packages/Sort.chpl [hunk 28]
```diff
@@ -18,53 +18,321 @@
...
+   algorithm.
+
+   :arg x: An iterable value to be sorted and yielded element by element
+   :type x: `iterable`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+   :yields: The elements of x in sorted order
+   :ytype: x's element type
+
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I'd add a .. note here that it's a TODO to make a standalone or leader/follower parallel iterator here.

Comment (mppf) [inline modules/packages/Sort.chpl]: ie. that the current version is only a serial iterator.

File: modules/packages/Sort.chpl [hunk 29]
```diff
@@ -283,64 +551,254 @@ proc QuickSort(Data: [?Dom] ?eltType, minlen=16, doublecheck=false, param revers
...
+  proc compare(a:?t, b) {
+    if a < b { return -1; }
+    else if b < a { return 1; }
+    else return 0;
+  }
+}
 
-  if (doublecheck) then VerifySort(Data, "SelectionSort", reverse);
+/* Reverse comparator built from another comparator.*/
+record ReverseComparator {
```
Comment (mppf) [inline modules/packages/Sort.chpl]: I wonder if this should just be called 'Reverse' ?

**Comment by mppf** [comment]
@ben-albrecht - looks great. I had lots of little feedback, mostly about comments and names. Please let's discuss individual naming suggestions if it seems we are not in agreement.

Generally, I think there are some Future Work steps that I'd like you to call out in the PR/merge message:

* implement radix sort
* move individual sort algorithms to a SortAlgorithms module to reduce the complexity of Sort
   (since most uses of this library will only want to call 'sort', 'sorted' or 'isSorted').
* implement parallel sorting algorithms and have some reasonable way for users to request serial or parallel sorts.

Thanks!

File: modules/packages/Sort.chpl (hunk 29 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: I like the symmetry of `ReverseComparator` and `DefaultComparator`. I considered `Reverse` and `Default` with `reverse` and `default` instances, but these would conflict with variables in the deprecated functions and would rob users of some excellent variable names.

File: modules/packages/Sort.chpl (hunk 27 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: Original comment by @bradcray, so he can correct me if I'm off:

I think it is saying that it would preferably create a new array `y` which shares the same element types `x`, but this requires a size to be passed (or we could calculate the size on the fly). This would be valuable for supporting iterable types other than arrays, such as iterators and lists.

File: modules/packages/Sort.chpl (hunk 28 above)
Comment (ben-albrecht) [inline modules/packages/Sort.chpl]: `TODO` + `.. note` added

File: modules/packages/Sort.chpl (hunk 27 above)
Comment (bradcray) [inline modules/packages/Sort.chpl]: Sorry that this wasn't clear on its own -- it's been long enough that I don't recall exactly what I was asking for, but my interpretation today is that I wanted a way to assert "y is an array" (where currently 'y' isn't constrained to be anything), but that I didn't have any way to do so since I couldn't predict its size.  This is akin to the longstanding desire to be able to say things like:

    var A: [1..] real = ...;

as a means of saying "It should be a 1-based array, but its size should be taken from the initialization expression.  We've discussed before how perhaps we should be able to declare variables that are first-class captures of iterators (like the David Wonnacott mail to chapel-bugs this week), and if we had those semantics, the current code wouldn't necessarily result in 'y' being array -- instead, it might be considered a capture of 'x' in the event that 'x' was, say, an iterator (or a non-array data structure that supported iteration?).  

I think the part about the element type is saying that it'd be nice if you could query arbitrary iterable things to ask what type you'll get when you iterate over them, so that one could say something like:

    var y: [1..] x.yieldType = for xs in x do (xs + 1);
